src/Microsoft.AspNetCore.TestHost/ResponseStream.cs

```diff
@@ -93,126 +77,68 @@ public override void SetLength(long value)
+            cancellationToken.ThrowIfCancellationRequested();
             CheckNotComplete();
 
-            _writeLock.Wait();
+            await _writeLock.WaitAsync(cancellationToken);
             try
             {
-                FirstWriteAsync().GetAwaiter().GetResult();
+                await FirstWriteAsync();
+                await _pipe.Writer.FlushAsync(cancellationToken);
             }
             finally
             {
                 _writeLock.Release();
             }
-
-            // TODO: Wait for data to drain?
         }
 
-        public override Task FlushAsync(CancellationToken cancellationToken)
+        public override int Read(byte[] buffer, int offset, int count)
         {
-            if (cancellationToken.IsCancellationRequested)
-            {
-                TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();
-                tcs.TrySetCanceled();
-                return tcs.Task;
-            }
-
-            Flush();
-
-            // TODO: Wait for data to drain?
-
-            return Task.FromResult<object>(null);
+            return ReadAsync(buffer, offset, count, CancellationToken.None).GetAwaiter().GetResult();
         }
 
-        public override int Read(byte[] buffer, int offset, int count)
+        public async override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
             VerifyBuffer(buffer, offset, count, allowEmpty: false);
-            _readLock.Wait();
+            CheckAborted();
+            var registration = cancellationToken.Register(Abort);
             try
             {
-                int totalRead = 0;
-                do
+                // TODO: Usability issue. Flush or zero byte write causes ReadAsync to complete without data so I have to call ReadAsync in a loop.
+                while (true)
                 {
-                    // Don't drain buffered data when signaling an abort.
-                    CheckAborted();
-                    if (_topBuffer.Count <= 0)
+                    var result = await _pipe.Reader.ReadAsync(cancellationToken);
+
+                    var readableBuffer = result.Buffer;
+                    if (!readableBuffer.IsEmpty)
                     {
-                        byte[] topBuffer = null;
-                        while (!_bufferedData.TryDequeue(out topBuffer))
-                        {
-                            if (_complete)
-                            {
-                                CheckAborted();
-                                // Graceful close
-                                return totalRead;
-                            }
-                            WaitForDataAsync().Wait();
-                        }
-                        _topBuffer = new ArraySegment<byte>(topBuffer);
+                        var actual = Math.Min(readableBuffer.Length, count);
+                        readableBuffer = readableBuffer.Slice(0, actual);
```

Comment:
This is a bug in both of our loops but you don't use the offset parameter at all.

Reply:
Fixed, see the next line.